src/coreclr/jit/importercalls.cpp

```diff
@@ -3093,6 +3094,29 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Type_get_TypeHandle:
+            {
+                assert(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+
+                // Try to avoid ClsHandle -> Type object -> ClsHandle roundtrip:
+                GenTree* op1 = impStackTop(0).val;
+                if (op1->IsHelperCall() && gtIsTypeHandleToRuntimeTypeHelper(op1->AsCall()))
+                {
+                    // struct RuntimeTypeHandle { IntPtr _value; }
+                    assert(info.compCompHnd->getClassNumInstanceFields(sig->retTypeClass) == 1);
+
+                    unsigned structLcl = lvaGrabTemp(true DEBUGARG("RuntimeTypeHandle"));
+                    lvaSetStruct(structLcl, sig->retTypeClass, false);
+                    GenTree*       realHandle   = op1->AsCall()->gtArgs.GetUserArgByIndex(0)->GetNode();
+                    GenTreeLclFld* handleFld    = gtNewLclFldNode(structLcl, realHandle->TypeGet(), 0);
+                    GenTree*       asgHandleFld = gtNewAssignNode(handleFld, realHandle);
```

Comment:
Would it be better to create this as `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE` helper call that takes the first arg of `op1` as an input?

We have other places in the JIT that look for `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE`. I do not think they will be able to recognize this.

Reply:
It probably makes sense but we'll need to fold it back to a constant somewhere later, I tried to do this and hit no diffs so maybe we better teach jit to recognize raw handles if we ever find a use case?

Reply:
The canonical shape for creating RuntimeTypeHandle from raw handle used in other places is `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE` helper. I think we should stick to canonical shapes where possible, to allow pattern matching optimizations to compose.

Cleaning up `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE` in later stages applies to all places that use canonical shape for creating RuntimeTypeHandle from raw handle.

Reply:
Here is the change you wanted to see https://github.com/EgorBo/runtime-1/commit/55279bc8b41abd31b82997b5083bd6c0f0b868be

It seems to be not needed for CoreCLR since we have to wrap RuntimeType object and I've ran SPMI searching for `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE(class handle)` and found 0 cases.

I spent some time trying to find a case where this helps and wasn't able to do so, I only see a worse codegen (a stack spill I can't explain) - my understanding that we promote structs right after importer so it's better to expand them there.

Reply:
I believe that the right way to do this optimization is without any AOT special-casing and without `RuntimeType` special-casing in `getObjectContent`. I think you are saying that it is complicated and not worth it, and that you would prefer to either do nothing or do some of the special casing.

@dotnet/jit-contrib Any opinions about this?

Reply:
Right, I just feel that the energy is better to spent elsewhere since:
1) Does NativeAOT needs this optimization? It doesn't look so since GC.NativeAOT.cs doesn't use it (although. it's fully implemented in this PR including shared generics)
2) Is this pattern popular? Only one hit in `GC.CoreCLR.cs` in our repo + a few hits in OSS judging by grep.app

So we only miss the shared generic case on CoreCLR and unloadable ALC (on CoreCLR). To support both of them we need to mark two methods with [Intrinsic] (`Type.get_TypeHandle` and `RuntimeTypeHandle.get_Value` (before it is inlined) - in this case we also will need some hacky way to work with a spilled struct `RuntimeTypeHandle` before we can detect the shape from two intrinsics.
```
STMT00000 ( 0x000[E-] ... 0x00F )
               [000003] I-C-G------                         *  CALL      struct System.RuntimeType:get_TypeHandle():System.RuntimeTypeHandle:this (exactContextHnd=0x00007FF7A7BA11C9)
               [000002] --C-G------ this                    \--*  CALL help ref    CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
               [000000] H---------- arg0                       \--*  CNS_INT(h) long   0x7ff7a7cfa2a8 class

***** BB01
STMT00001 ( 0x000[E-] ... ??? )
               [000006] -AC--------                         *  ASG       struct (copy)
               [000005] D------N---                         +--*  LCL_VAR   struct<System.RuntimeTypeHandle, 8> V01 loc0         
               [000004] --C--------                         \--*  RET_EXPR  struct(for [000003])

***** BB01
STMT00002 ( 0x010[E-] ... 0x017 )
               [000008] I-C-G------                         *  CALL      long   System.RuntimeTypeHandle:get_Value():long:this (exactContextHnd=0x00007FF7A7D21919)
               [000007] ----------- this                    \--*  LCL_ADDR  byref  V01 loc0         [+0]

***** BB01
STMT00003 ( 0x010[E-] ... ??? )
               [000010] --C--------                         *  RETURN    long  
               [000009] --C--------                         \--*  RET_EXPR  long  (for [000008])
```
which is not trivial in importer

Reply:
> in this case we also will need some hacky way to work with a spilled struct RuntimeTypeHandle  

If it helps, I would be ok with ignoring `RuntimeTypeHandle.get_Value` and tell people to use `RuntimeTypeHandle.ToIntPtr` for best results. We have added `RuntimeTypeHandle.ToIntPtr/FromIntPtr` family of APIs recently, as a managed equivalent of some Mono embedding APIs.

Reply:
> > in this case we also will need some hacky way to work with a spilled struct RuntimeTypeHandle
> 
> If it helps, I would be ok with ignoring `RuntimeTypeHandle.get_Value` and tell people to use `RuntimeTypeHandle.ToIntPtr` for best results. We have added `RuntimeTypeHandle.ToIntPtr/FromIntPtr` family of APIs recently, as a managed equivalent of some Mono embedding APIs.

Yes, that one is at least do-able in importer, I've just merged it

Reply:
There is some special spill avoidance in the importer to make recognizing these kinds of multi-call patterns easier; you could extend this if needed.

https://github.com/dotnet/runtime/blob/7d54c0524da14337c916c754061a2a300fdfabb6/src/coreclr/jit/importercalls.cpp#L1414-L1442